File: vignettes/05_structured_output/csharp/05_structured_output.cs
// Structured Output — Getting typed responses from the LLM (C#)
//
// This sample demonstrates how to request structured JSON output from an
// agent, so you get validated C# objects instead of free-form text.
// It mirrors the Julia vignette 05_structured_output.
//
// Prerequisites:
//   - Ollama running locally with qwen3:8b pulled
//   - dotnet restore

using System.ComponentModel;
using System.Text.Json;
using System.Text.Json.Serialization;
using Microsoft.Agents.AI;
using Microsoft.Extensions.AI;
using OllamaSharp;

var endpoint = Environment.GetEnvironmentVariable("OLLAMA_ENDPOINT")
    ?? "http://localhost:11434";
var modelName = Environment.GetEnvironmentVariable("OLLAMA_MODEL_NAME")
    ?? "qwen3:8b";

// Create an agent backed by the local Ollama instance.
AIAgent agent = new OllamaApiClient(new Uri(endpoint), modelName)
    .AsAIAgent(
        instructions: "You are a movie critic. Return structured reviews when asked.",
        name: "MovieCritic");

// ── Structured output with ResponseFormat ────────────────────────────────
Console.WriteLine("=== Movie Review (structured) ===");

// Use RunAsync<T> for type-safe structured output.
AgentResponse<MovieReview> response = await agent.RunAsync<MovieReview>(
    "Review the movie 'Inception' (2010).");

MovieReview review = response.Result;
Console.WriteLine($"Title: {review.Title}");
Console.WriteLine($"Rating: {review.Rating}");
Console.WriteLine($"Summary: {review.Summary}");
Console.WriteLine($"Recommended: {review.Recommended}");
Console.WriteLine();

// ── Manual approach: deserialize from response text ──────────────────────
Console.WriteLine("=== Movie Review (manual parse) ===");

// Create agent with explicit response format in options.
AIAgent agentWithFormat = new OllamaApiClient(new Uri(endpoint), modelName)
    .AsAIAgent(new ChatClientAgentOptions()
    {
        Name = "MovieCritic",
        ChatOptions = new()
        {
            Instructions = "You are a movie critic. Return structured reviews when asked.",
            ResponseFormat = ChatResponseFormat.ForJsonSchema<MovieReview>()
        }
    });

AgentResponse rawResponse = await agentWithFormat.RunAsync(
    "Review 'The Shawshank Redemption'.");

Console.WriteLine($"Raw: {rawResponse.Text}");
MovieReview parsed = JsonSerializer.Deserialize<MovieReview>(rawResponse.Text)!;
Console.WriteLine($"Title: {parsed.Title}");
Console.WriteLine($"Rating: {parsed.Rating}");
Console.WriteLine();

// ── Nested types ─────────────────────────────────────────────────────────
Console.WriteLine("=== Book Review (nested types) ===");

AgentResponse<BookReview> bookResponse = await agent.RunAsync<BookReview>(
    "Review '1984' by George Orwell.");

BookReview book = bookResponse.Result;
Console.WriteLine($"Title: {book.Title}");
Console.WriteLine($"Author: {book.Author.Name} ({book.Author.Nationality})");
Console.WriteLine($"Rating: {book.Rating}");
Console.WriteLine($"Themes: {string.Join(", ", book.Themes)}");

// ── Response type definitions ────────────────────────────────────────────

[Description("A structured movie review")]
public sealed class MovieReview
{
    [JsonPropertyName("title")]
    public string Title { get; set; } = "";

    [JsonPropertyName("rating")]
    public int Rating { get; set; }

    [JsonPropertyName("summary")]
    public string Summary { get; set; } = "";

    [JsonPropertyName("recommended")]
    public bool Recommended { get; set; }
}

[Description("Information about an author")]
public sealed class Author
{
    [JsonPropertyName("name")]
    public string Name { get; set; } = "";

    [JsonPropertyName("nationality")]
    public string Nationality { get; set; } = "";
}

[Description("A structured book review")]
public sealed class BookReview
{
    [JsonPropertyName("title")]
    public string Title { get; set; } = "";

    [JsonPropertyName("author")]
    public Author Author { get; set; } = new();

    [JsonPropertyName("rating")]
    public int Rating { get; set; }

    [JsonPropertyName("themes")]
    public List<string> Themes { get; set; } = [];
}

File: vignettes/05_structured_output/python/05_structured_output.py
"""
Structured Output — Getting typed responses from the LLM (Python)

This sample demonstrates how to request structured JSON output from an
agent using Pydantic models, so you get validated Python objects instead
of free-form text. It mirrors the Julia vignette 05_structured_output.

Prerequisites:
  - Ollama running locally with `qwen3:8b` pulled
  - pip install agent-framework-ollama pydantic
"""

import asyncio
import json

from pydantic import BaseModel

from agent_framework.ollama import OllamaChatClient


# ── Define response types as Pydantic models ─────────────────────────────

class MovieReview(BaseModel):
    title: str
    rating: int
    summary: str
    recommended: bool


class Author(BaseModel):
    name: str
    nationality: str


class BookReview(BaseModel):
    title: str
    author: Author
    rating: int
    themes: list[str]


async def main() -> None:
    # Create a chat client pointing at the local Ollama instance.
    client = OllamaChatClient(
        host="http://localhost:11434",
        model_id="qwen3:8b",
    )

    # Ollama's structured output uses a JSON schema dict as `format`,
    # not a Pydantic class directly.  We embed the schema in the agent
    # instructions so the LLM knows the expected shape and use
    # format=<json-schema> to enforce valid JSON.

    movie_schema = MovieReview.model_json_schema()
    agent = client.as_agent(
        name="MovieCritic",
        instructions=(
            "You are a movie critic. Always return structured reviews as JSON "
            f"matching this schema:\n{json.dumps(movie_schema, indent=2)}"
        ),
    )

    # ── Typed response using response_format ─────────────────────────────
    print("=== Movie Review (structured) ===")
    result = await agent.run(
        "Review the movie 'Inception' (2010).",
        options={"response_format": "json"},
    )
    print(f"Raw response: {result.text}")

    # Parse the JSON response into a Pydantic model.
    review = MovieReview.model_validate_json(result.text)
    print(f"Title: {review.title}")
    print(f"Rating: {review.rating}")
    print(f"Summary: {review.summary}")
    print(f"Recommended: {review.recommended}")
    print()

    # ── Nested types ─────────────────────────────────────────────────────
    print("=== Book Review (nested types) ===")
    book_schema = BookReview.model_json_schema()
    book_agent = client.as_agent(
        name="BookCritic",
        instructions=(
            "You are a book critic. Always return structured reviews as JSON "
            f"matching this schema:\n{json.dumps(book_schema, indent=2)}"
        ),
    )
    result = await book_agent.run(
        "Review '1984' by George Orwell.",
        options={"response_format": "json"},
    )
    book = BookReview.model_validate_json(result.text)
    print(f"Title: {book.title}")
    print(f"Author: {book.author.name} ({book.author.nationality})")
    print(f"Rating: {book.rating}")
    print(f"Themes: {book.themes}")
    print()

    # ── Inspect the generated schema ─────────────────────────────────────
    print("=== Generated JSON Schema ===")
    print(json.dumps(MovieReview.model_json_schema(), indent=2))


if __name__ == "__main__":
    asyncio.run(main())
